### ISSUE TITLE
Autonomous Mode

### ISSUE THREAD

**Issue by gnawinganimal** [SouthEugeneRoboticsTeam/BunnyBots-2019#5]
Of course, our robot needs a sweet auto. It will probably use a combination of machine learning and [pathfinding](https://github.com/JacisNonsense/Pathfinder).

**Comment by andrewda** [comment]
A big issue with Pathfinder (and why we used meanlib's path following last year) is that it takes a _loooong_ time to compile paths. Like on the order of 3-5 seconds (1/3 of auto). Thus it's not feasible to dynamically update paths as you're moving. That's fine if you're certain about where you want to go from the beginning (e.g. trying to get to the cargo ship from a specific starting location), but once you introduce computer vision and ML, your idea of where you want to be will be a lot more accurate from closer up, so you'll need to be updating the path.

Now [Pathfinder v2](https://github.com/GrappleRobotics/Pathfinder) claims to have this capability, but it also hasn't been updated in a year and there's no telling whether it works at the moment. If you're planning on doing CV/ML + path following, there's a good chance you're going to have to implement something like meanlib's yourself (or... just literally use meanlib -- there's a good chance the math is abstracted far enough away from meanlib core that it can be used with Sertain).

**Comment by andrewda** [comment]
Now if you want to get _really_ crazy (and be like the Zebracorns), you can use [ROS path planning](http://wiki.ros.org/global_planner) which opens up a world of possibilities. I wish I had worked more on getting ROS to work on our robots since I fully expect it to be the future of FRC robots... But I expect that to take well over the month or two y'all have before BunnyBots 😝 

&nbsp;

&nbsp;

But still think how cool this would be, avoiding opposing robots as they darted in front of you:

![](http://mrsdprojects.ri.cmu.edu/2016teamd/wp-content/uploads/sites/15/2016/11/teb_straight_plan.gif)

**Comment by gnawinganimal** [comment]
It would be very good. I honestly don't know what path planning we are going to use. If I can't finish mine on time I guess we can use `TrajectoryLib` by the cheezypoofs.

**Comment by gnawinganimal** [comment]
`sertain-trajectories` is going pretty good, though. I think I can finish by bunnybots. I'm using `TrajectoryLib` as an example. I now just need to translate the spline into a trajectory for each wheel

**Comment by gnawinganimal** [comment]
`ROS` looks really dope. Honestly thats what `sertain` would be based off if I had the time

**Comment by andrewda** [comment]
@Jamdan2 yeah, ROS is great. It's essentially what everyone outside of FRC uses (including OSU Robotics Team 😜)
